api/middleware/rateLimiter.go

```diff
@@ -0,0 +1,106 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/api/shared"
+	"github.com/gin-gonic/gin"
+)
+
+type rateLimiter struct {
+	requestsMap    map[string]uint64
+	mutRequestsMap sync.RWMutex
+	limits         map[string]uint64
+	mutLimits      sync.RWMutex
+	countDuration  uint64
+}
+
+// NewRateLimiter returns a new instance of rateLimiter
+func NewRateLimiter(limits map[string]uint64, countDuration uint64) (*rateLimiter, error) {
+	if limits == nil {
+		return nil, ErrNilLimitsMapForEndpoints
+	}
+	return &rateLimiter{
+		requestsMap:   make(map[string]uint64),
+		limits:        limits,
+		countDuration: countDuration,
+	}, nil
+}
+
+// MiddlewareHandlerFunc returns the gin middleware for limiting the number of requests for a given endpoint
+func (rl *rateLimiter) MiddlewareHandlerFunc() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		endpoint := c.FullPath()
+
+		limitForEndpoint, isEndpointLimited := rl.getEndpointLimit(endpoint)
```

Comment:
I know this works as it is because we are only reading from the limits map but if we change this code, we might end up in a panic during runtime (there are no tests for concurrent MiddlewareHandlerFunc calls) Also. what happens with the rl.mutLimits field?

Reply:
removed the unused `mutLimits`. regarding the map reading, I would leave it as it is now, and if we'll have another use for the map, we could add the mutex protection then.